### PR TITLE
Add AV1 Decoder

### PR DESCRIPTION
**Pull request by vasn1k**
Adds support for the AV1 decoder

### CODE REVIEW

File: common/include/VkVideoCore/VulkanVideoCapabilities.h [hunk 1]
```diff
@@ -37,10 +37,15 @@ class VulkanVideoCapabilities
         videoCapabilities       =       VkVideoCapabilitiesKHR { VK_STRUCTURE_TYPE_VIDEO_CAPABILITIES_KHR, &videoDecodeCapabilities };
         VkVideoDecodeH264CapabilitiesKHR h264Capabilities    = { VK_STRUCTURE_TYPE_VIDEO_DECODE_H264_CAPABILITIES_KHR, nullptr };
         VkVideoDecodeH265CapabilitiesKHR h265Capabilities    = { VK_STRUCTURE_TYPE_VIDEO_DECODE_H265_CAPABILITIES_KHR, nullptr };
+        VkVideoDecodeAV1CapabilitiesKHR av1Capabilities{};
```
Comment (zlatinski) [inline common/include/VkVideoCore/VulkanVideoCapabilities.h]: Can we please use the same structure init pattern as above, for consistency?

File: common/libs/VkCodecUtils/VulkanVideoProcessor.cpp [hunk 2]
```diff
@@ -517,22 +518,68 @@ size_t VulkanVideoProcessor::ConvertFrameToNv12(VulkanDecodedFrame* pFrame,
...
 
     // Copy the chroma plane(s)
-    for (uint32_t plane = numCompatiblePlanes; plane < numPlanes; plane++) {
-        uint32_t srcPlane = std::min(plane, mpInfo->planesLayout.numberOfExtraPlanes);
-        uint8_t* pDst = pOutBuffer + yuvPlaneLayouts[plane].offset;
-        for (int height = 0; height < secondaryPlaneHeight; height++) {
-            const uint8_t* pSrc;
-            if (srcPlane != plane) {
-                pSrc = readImagePtr + layouts[srcPlane].offset + ((plane - 1) * bytesPerPixel) + (layouts[srcPlane].rowPitch * height);
+    if (mpInfo->planesLayout.bpp == YCBCRA_10BPP) {
```
Comment (zlatinski) [inline common/libs/VkCodecUtils/VulkanVideoProcessor.cpp]: Why do we need these changes?
We are converting the number of planes or layouts here, which may create  a lot of confusion.
This code is also slower. There are external utilities to convert from one layout to another.

File: common/libs/VkCodecUtils/VulkanVideoSession.cpp [hunk 3]
```diff
@@ -40,14 +42,15 @@ VkResult VulkanVideoSession::Create(const VulkanDeviceContext* vkDevCtx,
...
+    static const VkExtensionProperties av1StdExtensionVersion = { VK_STD_VULKAN_VIDEO_CODEC_AV1_DECODE_EXTENSION_NAME, VK_STD_VULKAN_VIDEO_CODEC_AV1_DECODE_SPEC_VERSION };
 
     VkVideoSessionCreateInfoKHR& createInfo = pNewVideoSession->m_createInfo;
     createInfo.flags = sessionCreateFlags;
     createInfo.pVideoProfile = pVideoProfile->GetProfile();
     createInfo.queueFamilyIndex = videoQueueFamily;
     createInfo.pictureFormat = pictureFormat;
     createInfo.maxCodedExtent = maxCodedExtent;
-    createInfo.maxDpbSlots = maxDpbSlots;
+    createInfo.maxDpbSlots =  maxDpbSlots;
```
Comment (zlatinski) [inline common/libs/VkCodecUtils/VulkanVideoSession.cpp]: Why do we need this formatting change?

File: common/include/VkVideoCore/VkVideoCoreProfile.h [hunk 4]
```diff
@@ -86,6 +89,22 @@ class VkVideoCoreProfile
...
+            if (pProfileExt && (pProfileExt->sType != VK_STRUCTURE_TYPE_VIDEO_DECODE_AV1_PROFILE_INFO_KHR)) {
+                m_profile.sType = VK_STRUCTURE_TYPE_APPLICATION_INFO;
+                return false;
+            }
+            if (pProfileExt) {
+                m_av1DecodeProfile = *pProfileExt;
+            } else {
+                //  Use default ext profile parameters
+                m_av1DecodeProfile.sType         = VK_STRUCTURE_TYPE_VIDEO_DECODE_H265_PROFILE_INFO_KHR;
+                m_av1DecodeProfile.stdProfile = STD_VIDEO_AV1_PROFILE_MAIN; // @review: why are we hardcoding profiles here, they should come from the parser.
```
Comment (zlatinski) [inline common/include/VkVideoCore/VkVideoCoreProfile.h]: It is when the codec profile is not specified, which should be a rare case.

File: common/libs/VkCodecUtils/VulkanVideoProcessor.cpp [hunk 5]
```diff
@@ -145,6 +145,7 @@ int32_t VulkanVideoProcessor::Initialize(const VulkanDeviceContext* vkDevCtx,
                                     m_videoStreamDemuxer->GetChromaBitDepth(),
                                     m_videoStreamDemuxer->GetProfileIdc());
 
+
```
Comment (zlatinski) [inline common/libs/VkCodecUtils/VulkanVideoProcessor.cpp]: Extra line above for no reason?

File: common/libs/VkCodecUtils/VulkanVideoProcessor.cpp [hunk 6]
```diff
@@ -552,12 +612,23 @@ size_t VulkanVideoProcessor::OutputFrameToFile(VulkanDecodedFrame* pFrame)
         return (size_t)-1;
     }
 
+    assert(pFrame != nullptr);
+
     VkSharedBaseObj<VkImageResourceView> imageResourceView;
-    pFrame->imageViews[DecodeFrameBufferIf::IMAGE_TYPE_IDX_OUT].GetImageResourceView(imageResourceView);
+    pFrame->imageViews[DecodeFrameBufferIf::IMAGE_TYPE_IDX_FILTER].GetImageResourceView(imageResourceView);
+    // TODO: It would be nice if the index for the linear image did not migrate.
+    // With AV1 + Filmgrain it's questionable whether we want the linear image to be in IDX_FILTER or in IDX_OUT
```
Comment (zlatinski) [inline common/libs/VkCodecUtils/VulkanVideoProcessor.cpp]: The linear image should always be the result of the output.
If the output does not contain FG, then the linear image should be based on the decoded image.
If the output contains FG, then  the linear image should be based on the FG image.

File: common/libs/VkCodecUtils/VulkanVideoSession.cpp [hunk 7]
```diff
@@ -23,6 +23,8 @@
 #include "VkVideoCore/VkVideoCoreProfile.h"
 #include "VkCodecUtils/VulkanVideoSession.h"
 
+static constexpr bool DEBUGGING = false;
```
Comment (zlatinski) [inline common/libs/VkCodecUtils/VulkanVideoSession.cpp]: Please remove this above.

File: vk_video_decoder/demos/vk-video-dec/CMakeLists.txt [hunk 8]
```diff
@@ -142,15 +142,11 @@ elseif(${CMAKE_SYSTEM_NAME} STREQUAL "Linux")
     list(APPEND libraries PRIVATE -ldl -lrt -lpthread)
 
     if(BUILD_WSI_XCB_SUPPORT AND DEMOS_WSI_SELECTION STREQUAL "XCB")
-        find_package(XCB REQUIRED)
```
Comment (zlatinski) [inline vk_video_decoder/demos/vk-video-dec/CMakeLists.txt]: Why were these removed?

File: vk_video_decoder/demos/vk-video-dec/CMakeLists.txt [hunk 9]
```diff
@@ -142,15 +142,11 @@ elseif(${CMAKE_SYSTEM_NAME} STREQUAL "Linux")
...
-
         list(APPEND sources ${VK_VIDEO_COMMON_LIBS_SOURCE_ROOT}/VkShell/ShellXcb.cpp ${VK_VIDEO_COMMON_LIBS_SOURCE_ROOT}/VkShell/ShellXcb.h)
         list(APPEND definitions PRIVATE -DVK_USE_PLATFORM_XCB_KHR)
         list(APPEND includes PRIVATE ${XCB_INCLUDES})
         list(APPEND libraries PRIVATE ${XCB_LIBRARIES})
     elseif(BUILD_WSI_WAYLAND_SUPPORT AND DEMOS_WSI_SELECTION STREQUAL "WAYLAND")
-        find_package(Wayland REQUIRED)
```
Comment (zlatinski) [inline vk_video_decoder/demos/vk-video-dec/CMakeLists.txt]: Why were these removed?

File: vk_video_decoder/demos/vk-video-dec/CMakeLists.txt [hunk 10]
```diff
@@ -174,4 +170,7 @@ target_include_directories(vk-video-dec-test ${includes})
 target_link_libraries(vk-video-dec-test ${libraries})
 add_dependencies(vk-video-dec-test generate_helper_files)
 
-install(TARGETS vk-video-dec-test RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
\ No newline at end of file
+install(TARGETS vk-video-dec-test RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
+
+include(CTest)
```
Comment (zlatinski) [inline vk_video_decoder/demos/vk-video-dec/CMakeLists.txt]: What is this above for?

File: vk_video_decoder/include/vkvideo_parser/PictureBufferBase.h [hunk 11]
```diff
@@ -26,10 +26,15 @@ class VkPicIf {
 public:
     virtual void AddRef() = 0;
     virtual void Release() = 0;
-    int32_t decodeWidth;
-    int32_t decodeHeight;
-    int32_t decodeSuperResWidth;
-    int32_t reserved[16 - 3];
+
+    // This is only used for AV1, need because of 5.9.7
+    int32_t upscaledWidth;
```
Comment (zlatinski) [inline vk_video_decoder/include/vkvideo_parser/PictureBufferBase.h]: Why do we need these fields here, as part of the VkPicIf?
Is this something we need to communicate back to the parser?

File: vk_video_decoder/include/vkvideo_parser/VulkanVideoParserIf.h [hunk 12]
```diff
@@ -21,8 +21,13 @@
 #include "VkCodecUtils/VkVideoRefCountBase.h"
 #include "vkvideo_parser/StdVideoPictureParametersSet.h"
 #include "VkCodecUtils/VulkanBitstreamBuffer.h"
-#include "vk_video/vulkan_video_codecs_common.h"
 
+#ifndef DE_BUILD_VIDEO
+    #include "vk_video/vulkan_video_codecs_common.h"
```
Comment (zlatinski) [inline vk_video_decoder/include/vkvideo_parser/VulkanVideoParserIf.h]: Is this above for CTS? If yes, maybe a note is required above.

File: vk_video_decoder/include/vkvideo_parser/VulkanVideoParserIf.h [hunk 13]
```diff
@@ -324,188 +330,52 @@ typedef struct VkParserVp9PictureData {
...
+	uint16_t MiColStarts[64];
+	uint16_t MiRowStarts[64];
+	uint16_t width_in_sbs_minus_1[64];
+	uint16_t height_in_sbs_minus_1[64];
+	// --- End of tileInfo data ---
+
+	StdVideoAV1Quantization			quantization;
+	StdVideoAV1Segmentation			segmentation;
+	StdVideoAV1LoopFilter			loopFilter;
+	StdVideoAV1CDEF					CDEF;
```
Comment (zlatinski) [inline vk_video_decoder/include/vkvideo_parser/VulkanVideoParserIf.h]: Filed alignment of CDEF. Also, why not call this cdef (all small letters)?

File: vk_video_decoder/include/vkvideo_parser/VulkanVideoParserIf.h [hunk 14]
```diff
@@ -324,188 +330,52 @@ typedef struct VkParserVp9PictureData {
...
+	StdVideoAV1LoopRestoration		loopRestoration;
+	StdVideoAV1GlobalMotion			globalMotion;
+	StdVideoAV1FilmGrain			filmGrain;
```
Comment (zlatinski) [inline vk_video_decoder/include/vkvideo_parser/VulkanVideoParserIf.h]: tileOffsets[] and tileSizes[] are misaligned.

File: vk_video_decoder/include/vkvideo_parser/VulkanVideoParserIf.h [hunk 15]
```diff
@@ -524,16 +394,15 @@ typedef struct VkParserPictureData {
...
     int32_t chroma_format; // Chroma Format (should match sequence info)
     int32_t picture_order_count; // picture order count (if known)
-    uint8_t* pSideData; // Encryption Info
-    uint32_t sideDataLen; // Encryption Info length
 
     // Codec-specific data
     union {
         VkParserH264PictureData h264;
         VkParserHevcPictureData hevc;
+		VkParserAv1PictureData av1;
```
Comment (zlatinski) [inline vk_video_decoder/include/vkvideo_parser/VulkanVideoParserIf.h]: VkParserAv1PictureData av1 is misaligned.

File: vk_video_decoder/include/vkvideo_parser/VulkanVideoParserIf.h [hunk 16]
```diff
@@ -626,7 +494,7 @@ class VkParserVideoDecodeClient {
         = 0; // Returns max number of reference frames
         // (always at least 2 for MPEG-2)
     virtual bool
-    AllocPictureBuffer(VkPicIf** ppPicBuf)
+    AllocPictureBuffer(VkPicIf** ppPicBuf, uint32_t codedWidth, uint32_t codedHeight)
```
Comment (zlatinski) [inline vk_video_decoder/include/vkvideo_parser/VulkanVideoParserIf.h]: I see why we need these fields, but I don't think this information is needed. Images are allocated with the maximum size required, and they will not get allocated on this request. I don't think we need to change this prototype.

File: vk_video_decoder/include/vkvideo_parser/VulkanVideoParserIf.h [hunk 17]
```diff
@@ -645,7 +513,7 @@ class VkParserVideoDecodeClient {
         const uint8_t* pbData, size_t cbData)
         = 0; // Called for custom NAL parsing (not required)
     virtual uint32_t GetDecodeCaps() { return 0; } // NVD_CAPS_XXX
-    virtual int32_t GetOperatingPoint(VkParserOperatingPointInfo* pOPInfo)
+    virtual int32_t GetOperatingPoint(VkParserOperatingPointInfo*)
```
Comment (zlatinski) [inline vk_video_decoder/include/vkvideo_parser/VulkanVideoParserIf.h]: Do we really need that function?

File: vk_video_decoder/include/vkvideo_parser/VulkanVideoParserParams.h [hunk 18]
```diff
@@ -34,12 +34,15 @@ struct VkParserPerFrameDecodeParameters {
...
     int currPicIdx; /** Output index of the current picture                       */
+
     // VPS
     const StdVideoPictureParametersSet*     pStdVps;
     // SPS
     const StdVideoPictureParametersSet*     pStdSps;
     // PPS
     const StdVideoPictureParametersSet*     pStdPps;
+    // AV1 sequence parameter set
+    const StdVideoPictureParametersSet*     pStdAv1Sps;
```
Comment (zlatinski) [inline vk_video_decoder/include/vkvideo_parser/VulkanVideoParserParams.h]: I don't think we need this extra field for AV1. We can use pStdSps for AV1.

File: vk_video_decoder/include/vkvideo_parser/VulkanVideoParserParams.h [hunk 19]
```diff
@@ -48,13 +51,16 @@ struct VkParserPerFrameDecodeParameters {
...
-    VkSharedBaseObj<VulkanBitstreamBuffer> bitstreamData; /** bitstream data for this picture (slice-layer) */
+    size_t bitstreamDataLen;   // Number of bytes in bitstream data buffer
+    VkSharedBaseObj<VulkanBitstreamBuffer> bitstreamData; // bitstream data for this picture (slice-layer)
     VkVideoDecodeInfoKHR decodeFrameInfo;
     VkVideoPictureResourceInfoKHR dpbSetupPictureResource;
     int32_t numGopReferenceSlots;
     int8_t pGopReferenceImagesIndexes[MAX_DPB_REF_AND_SETUP_SLOTS];
     VkVideoPictureResourceInfoKHR pictureResources[MAX_DPB_REF_AND_SETUP_SLOTS];
+
+	bool filmGrainEnabled;
```
Comment (zlatinski) [inline vk_video_decoder/include/vkvideo_parser/VulkanVideoParserParams.h]: filmGrainEnabled is misaligned.

File: vk_video_decoder/include/vkvideo_parser/VulkanVideoParserParams.h [hunk 20]
```diff
@@ -48,13 +51,16 @@ struct VkParserPerFrameDecodeParameters {
...
+    bool isAV1;
```
Comment (zlatinski) [inline vk_video_decoder/include/vkvideo_parser/VulkanVideoParserParams.h]: Don't we have access to the codec type here? Why do we need this field above?

File: vk_video_decoder/include/vkvideo_parser/VulkanVideoParserParams.h [hunk 21]
```diff
@@ -88,8 +94,8 @@ struct VkParserDecodePictureInfo {
     uint64_t decodePicCount;
     VkVideotimestamp timestamp; // decode time
     VkParserFrameSyncinfo frameSyncinfo;
-    uint16_t videoFrameType; // VideoFrameType - use Vulkan codec specific type pd->CodecSpecific.h264.slice_type.
     uint16_t viewId; // HEVC nuh_layer_id & from pictureInfoH264->ext.mvcext.view_id
+    bool filmGrainEnabled;
```
Comment (zlatinski) [inline vk_video_decoder/include/vkvideo_parser/VulkanVideoParserParams.h:93]: I don't understand why filmGrainEnabled has a place in FrameSyncInfo?

File: vk_video_decoder/libs/NvVideoParser/include/VulkanAV1Decoder.h [hunk 22]
```diff
@@ -0,0 +1,477 @@
...
+    bool                        m_bSPSReceived;
+    bool                        m_bSPSChanged;
+    bool                        m_obuAnnexB;
+    uint8_t                     timing_info_present;
+    av1_timing_info_t           timing_info;
+    av1_dec_model_info_t        buffer_model;
+    av1_dec_model_op_params_t   op_params[MAX_NUM_OPERATING_POINTS + 1];
+    uint32_t                    op_frame_timing[MAX_NUM_OPERATING_POINTS + 1];
+
+	uint8_t						delta_frame_id_length;
```
Comment (zlatinski) [inline vk_video_decoder/libs/NvVideoParser/include/VulkanAV1Decoder.h]: A lot of fields are misaligned here.  I think we should use astyle or clang-format against this file.

File: vk_video_decoder/libs/NvVideoParser/include/VulkanH264Decoder.h [hunk 23]
```diff
@@ -809,7 +818,7 @@ class VulkanH264Decoder: public VulkanVideoDecoder
     bool dpb_sequence_start(slice_header_s *slh);
     void dpb_picture_start(pic_parameter_set_s *pps, slice_header_s *slh);
     void dpb_picture_end();
-    VkPicIf *alloc_picture();
+    VkPicIf *alloc_picture(uint32_t width, uint32_t height);
```
Comment (zlatinski) [inline vk_video_decoder/libs/NvVideoParser/include/VulkanH264Decoder.h]: I don't think we need the dimensions with alloc_picture() here. There are no pictures allocated, but rather just a picture allocated from a image pool with a fixed image size.

File: vk_video_decoder/libs/NvVideoParser/src/VulkanAV1Decoder.cpp [hunk 24]
```diff
@@ -0,0 +1,2488 @@
...
+
+#include "VulkanAV1Decoder.h"
+
+#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof(arr[0]))
+
+// constructor
+VulkanAV1Decoder::VulkanAV1Decoder(VkVideoCodecOperationFlagBitsKHR std, bool annexB)
+    : VulkanVideoDecoder(std)
+{
+	memset(&m_PicData, 0, sizeof(m_PicData));
```
Comment (zlatinski) [inline vk_video_decoder/libs/NvVideoParser/src/VulkanAV1Decoder.cpp:42]: Again, this file has misaligned fields. Please use astyle tool against it.

File: vk_video_decoder/libs/NvVideoParser/src/VulkanAV1GlobalMotionDec.cpp [hunk 25]
```diff
@@ -0,0 +1,262 @@
...
+                               (ref_params->wmmat[1] >> trans_prec_diff)) *
+                               trans_dec_factor;
+    }
+
+    return 1;
+}
+
+uint32_t VulkanAV1Decoder::DecodeGlobalMotionParams()
+{
+	StdVideoDecodeAV1PictureInfo* pStd = &m_PicData.std_info;
```
Comment (zlatinski) [inline vk_video_decoder/libs/NvVideoParser/src/VulkanAV1GlobalMotionDec.cpp]: astyle tool against this file.

File: vk_video_decoder/libs/NvVideoParser/src/VulkanH264Parser.cpp [hunk 26]
```diff
@@ -566,7 +566,10 @@ bool VulkanH264Decoder::BeginPicture_SVC(VkParserPictureData *pnvpd)
                         m_ds->dpb_entry[16].pPicBufRefBase = NULL;
                     }
                     // allocate buffer for current frame
-                    m_ds->dpb_entry[16].pPicBuf = alloc_picture();
+					uint32_t width = (m_sps->pic_width_in_mbs_minus1 + 1) * 16;
```
Comment (zlatinski) [inline vk_video_decoder/libs/NvVideoParser/src/VulkanH264Parser.cpp]: Bad alignment here above and below.

File: vk_video_decoder/libs/NvVideoParser/src/VulkanH265Parser.cpp [hunk 27]
```diff
@@ -2719,7 +2720,16 @@ void VulkanH265Decoder::dpb_picture_start(VkSharedBaseObj<hevc_pic_param_s>& pps
     cur->output = PicOutputFlag;
     if (!cur->pPicBuf)
     {
-        if (!m_pClient->AllocPictureBuffer(&cur->pPicBuf))
+	    const hevc_seq_param_s* sps = m_spss[pps->pps_seq_parameter_set_id];
```
Comment (zlatinski) [inline vk_video_decoder/libs/NvVideoParser/src/VulkanH265Parser.cpp]: Bad alignment - use of tabs and spaces. Doesn't git complain about it?

File: vk_video_decoder/libs/VkDecoderUtils/FFmpegDemuxer.cpp [hunk 28]
```diff
@@ -308,6 +308,9 @@ class FFmpegDemuxer : public VideoStreamDemuxer {
     #ifdef VK_EXT_video_decode_vp9
         case AV_CODEC_ID_VP9        : return VK_VIDEO_CODEC_OPERATION_DECODE_VP9_BIT_KHR;
     #endif // VK_EXT_video_decode_vp9
+    #ifdef vulkan_video_codec_av1std_decode
+		case AV_CODEC_ID_AV1        : return VK_VIDEO_CODEC_OPERATION_DECODE_AV1_BIT_KHR;
```
Comment (zlatinski) [inline vk_video_decoder/libs/VkDecoderUtils/FFmpegDemuxer.cpp]: Misaligned fields. Somebody needs to fix their editor to use spaces and not tabs.

**Review by zlatinski — CHANGES_REQUESTED** [review]
Thanks for the integration, Vassili!
I'm at
vk_video_decoder/libs/VkVideoDecoder/VkParserVideoPictureParameters.cpp. Still more to go.

File: vk_video_decoder/libs/VulkanVideoFrameBuffer/VulkanVideoFrameBuffer.h [hunk 29]
```diff
@@ -69,19 +69,30 @@ class VulkanVideoFrameBuffer : public IVulkanVideoFrameBufferParserCb {
...
+        , pStdVps(pStdVpsRef)
+        , pStdAv1Pps(nullptr) {}
+
+        ReferencedObjectsInfo(const VkVideoRefCountBase* pBitstreamDataRef,
+                              const VkVideoRefCountBase* pStdAv1PpsRef)
+        : pBitstreamData(pBitstreamDataRef)
+        , pStdPps(nullptr)
+        , pStdSps(nullptr)
+        , pStdVps(nullptr)
+        , pStdAv1Pps(pStdAv1PpsRef) {}
```
Comment (zlatinski) [inline vk_video_decoder/libs/VulkanVideoFrameBuffer/VulkanVideoFrameBuffer.h]: We don't need pStdAv1Pps. Please use pStdSpsRef.

File: vk_video_decoder/libs/VulkanVideoFrameBuffer/VulkanVideoFrameBuffer.h [hunk 30]
```diff
@@ -69,19 +69,30 @@ class VulkanVideoFrameBuffer : public IVulkanVideoFrameBufferParserCb {
         // VPS
         const VkVideoRefCountBase*     pStdVps;
 
+        const VkVideoRefCountBase*     pStdAv1Pps;
```
Comment (zlatinski) [inline vk_video_decoder/libs/VulkanVideoFrameBuffer/VulkanVideoFrameBuffer.h]: We don't need pStdAv1Pps. Please use pStdSpsRef.

File: vk_video_decoder/libs/VulkanVideoFrameBuffer/VulkanVideoFrameBuffer.cpp [hunk 31]
```diff
@@ -157,6 +157,8 @@ class NvPerFrameDecodeResources : public vkPicBuffBase {
     VkSharedBaseObj<VkVideoRefCountBase>  stdSps;
     // PPS
     VkSharedBaseObj<VkVideoRefCountBase>  stdPps;
+    // AV1 SPS
+    VkSharedBaseObj<VkVideoRefCountBase>  stdAv1Pps;
```
Comment (zlatinski) [inline vk_video_decoder/libs/VulkanVideoFrameBuffer/VulkanVideoFrameBuffer.cpp]: We don't need pStdAv1Pps. Please use pStdSpsRef.

File: vk_video_decoder/libs/VulkanVideoFrameBuffer/VulkanVideoFrameBuffer.cpp [hunk 32]
```diff
@@ -838,6 +840,7 @@ void NvPerFrameDecodeResources::Deinit()
     stdPps = nullptr;
     stdSps = nullptr;
     stdVps = nullptr;
+    stdAv1Pps = nullptr;
```
Comment (zlatinski) [inline vk_video_decoder/libs/VulkanVideoFrameBuffer/VulkanVideoFrameBuffer.cpp]: We don't need pStdAv1Pps. Please use pStdSpsRef.

File: vk_video_decoder/libs/VkVideoDecoder/VkVideoDecoder.h [hunk 33]
```diff
@@ -220,13 +220,13 @@ class VkVideoDecoder : public IVulkanVideoDecoderHandler {
         , m_decodeFramesData(vkDevCtx)
         , m_decodePicCount(0)
         , m_hwLoadBalancingTimelineSemaphore()
-        , m_dpbAndOutputCoincide(true)
+        , m_preferDPBAndOutputCoincide(true)
```
Comment (zlatinski) [inline vk_video_decoder/libs/VkVideoDecoder/VkVideoDecoder.h]: m_preferDPBAndOutputCoincide - please don't change the name here.
There is nothing to prefer here. Based on the implementation report, the application must use a coincide mode or not.

File: vk_video_decoder/libs/VkVideoDecoder/VkVideoDecoder.h [hunk 34]
```diff
@@ -220,13 +220,13 @@ class VkVideoDecoder : public IVulkanVideoDecoderHandler {
...
         , m_videoMaintenance1FeaturesSupported(false)
         , m_enableDecodeFilter((enableDecoderFeatures & ENABLE_POST_PROCESS_FILTER) != 0)
         , m_useImageArray(false)
         , m_useImageViewArray(false)
-        , m_useSeparateOutputImages(((enableDecoderFeatures & ENABLE_LINEAR_OUTPUT) != 0) || m_enableDecodeFilter)
-        , m_useLinearOutput((enableDecoderFeatures & ENABLE_LINEAR_OUTPUT) != 0)
+        , m_useSeparateOutputImages(false) // TODO: Move to correct place this should not happen during class instantiation ((enableDecoderFeatures & ENABLE_LINEAR_OUTPUT) != 0) || m_enableDecodeFilter)
```
Comment (zlatinski) [inline vk_video_decoder/libs/VkVideoDecoder/VkVideoDecoder.h]: Why are the TODOs here? I don't get it.

File: vk_video_decoder/libs/VkVideoDecoder/VkVideoDecoder.h [hunk 35]
```diff
@@ -317,7 +317,7 @@ class VkVideoDecoder : public IVulkanVideoDecoderHandler {
     uint64_t                                         m_decodePicCount; // Also used for the HW load balancing timeline semaphore
     VkSharedBaseObj<VkParserVideoPictureParameters>  m_currentPictureParameters;
     VkSemaphore m_hwLoadBalancingTimelineSemaphore;
-    uint32_t m_dpbAndOutputCoincide : 1;
+    uint32_t m_preferDPBAndOutputCoincide : 1; // This is a preference, the code may still choose to split the DPB and Output. As in the case of AV1 filmgrain.
```
Comment (zlatinski) [inline vk_video_decoder/libs/VkVideoDecoder/VkVideoDecoder.h]: The separate output is an application choice based on linear/filtered vs. not, but not the coincide mode. (non-)coincide mode reported is mandatory for the applications.

File: vk_video_decoder/libs/VkVideoParser/VulkanVideoParser.cpp [hunk 36]
```diff
@@ -1544,6 +1653,8 @@ uint32_t VulkanVideoParser::FillDpbH265State(
         pStdPictureInfo->RefPicSetLtCurr[numPocLtCurr++] = 0xff;
     }
 
+    (void)numPocTotalCurr; // TODO
```
Comment (zlatinski) [inline vk_video_decoder/libs/VkVideoParser/VulkanVideoParser.cpp]: TODO what and why?

File: vk_video_decoder/libs/VkVideoDecoder/VkVideoDecoder.cpp [hunk 37]
```diff
@@ -231,19 +235,22 @@ int32_t VkVideoDecoder::StartVideoSequence(VkParserDetectedVideoFormat* pVideoFo
...
-    if (m_dpbAndOutputCoincide) {
+    bool filmGrainEnabled = ((videoCodec == VK_VIDEO_CODEC_OPERATION_DECODE_AV1_BIT_KHR) && (pVideoFormat->filmGrainUsed != false));
+    if ((m_preferDPBAndOutputCoincide != false) && (filmGrainEnabled == false)) {
         dpbImageUsage = outImageUsage | VK_IMAGE_USAGE_VIDEO_DECODE_DPB_BIT_KHR;
         outImageUsage &= ~VK_IMAGE_USAGE_VIDEO_DECODE_DST_BIT_KHR;
     } else {
         // The implementation does not support dpbAndOutputCoincide
         m_useSeparateOutputImages = true;
+        // When the implementation needs a separate output image, how to query whether the output image supports linear.
+        // Should the attempt of allocating an image with DECODE + LINEAR be attempted and if failed tried with just LINEAR?
```
Comment (zlatinski) [inline vk_video_decoder/libs/VkVideoDecoder/VkVideoDecoder.cpp]: Do you mean just Decode? Yes, it may make sense.

**Review by zlatinski — CHANGES_REQUESTED** [review]
I'm done with the second pass.
I'll do the third one after the previous issues are addressed.
Thank you, Vasilli!

File: common/libs/VkCodecUtils/VulkanVideoProcessor.cpp (hunk 2 above)
Comment (vasn1k) [reply common/libs/VkCodecUtils/VulkanVideoProcessor.cpp]: This was needed for VKCTS. I'll need to double check whether this is still the case.

File: vk_video_decoder/demos/vk-video-dec/CMakeLists.txt (hunk 10 above)
Comment (vasn1k) [reply vk_video_decoder/demos/vk-video-dec/CMakeLists.txt]: I assumed all this was to get CTS building. I'll remove it for now and when we do the merge for CTS we'll have to figure it out at that point.

File: vk_video_decoder/include/vkvideo_parser/PictureBufferBase.h (hunk 11 above)
Comment (vasn1k) [reply vk_video_decoder/include/vkvideo_parser/PictureBufferBase.h]: I changed them back.

File: vk_video_decoder/include/vkvideo_parser/VulkanVideoParserIf.h (hunk 13 above)
Comment (vasn1k) [reply vk_video_decoder/include/vkvideo_parser/VulkanVideoParserIf.h]: I've ran clang format on the entire file. CDEF is an acronym for Constrained Directional Enhancement Filter, and we are camel-ing our variables.

File: vk_video_decoder/include/vkvideo_parser/VulkanVideoParserIf.h (hunk 17 above)
Comment (vasn1k) [reply vk_video_decoder/include/vkvideo_parser/VulkanVideoParserIf.h]: gone.

File: vk_video_decoder/include/vkvideo_parser/VulkanVideoParserParams.h (hunk 21 above)
Comment (vasn1k) [reply vk_video_decoder/include/vkvideo_parser/VulkanVideoParserParams.h:93]: filmGrainEnabled is part of VkParserDecodePictureInfo

File: vk_video_decoder/libs/VkVideoParser/VulkanVideoParser.cpp (hunk 36 above)
Comment (vasn1k) [reply vk_video_decoder/libs/VkVideoParser/VulkanVideoParser.cpp]: This variable is unused, it needs to be removed or used.

**Comment by zlatinski** [comment]
pCurrFrameDecParams->decodeFrameInfo.srcBufferRange =  (pCurrFrameDecParams->bitstreamDataLen + 127) & ~127;

This change is not right.

See the spec:
VUID-vkCmdDecodeVideoKHR-pDecodeInfo-07139
pDecodeInfo->srcBufferRange must be an integer multiple of [VkVideoCapabilitiesKHR](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VkVideoCapabilitiesKHR)::minBitstreamBufferSizeAlignment, as returned by [vkGetPhysicalDeviceVideoCapabilitiesKHR](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#vkGetPhysicalDeviceVideoCapabilitiesKHR) for the video profile the bound video session was created with ...
 
Is the implementation reporting  minBitstreamBufferSizeAlignment = 128?

**Comment by vasn1k** [comment]
@zlatinski 
> minBitstreamBufferSizeAlignment
The implementation returns 256. 

>pCurrFrameDecParams->decodeFrameInfo.srcBufferRange = (pCurrFrameDecParams->bitstreamDataLen + 127) & ~127;
>This change is not right.

I wasn't entirely sure about this code and couldn't find more comments about it, it's interesting that it's aligning to a minimum of 128byte.

**Comment by zlatinski** [comment]
Already merged - thank you, Vassili!

File: vk_video_decoder/libs/NvVideoParser/src/VulkanAV1Decoder.cpp [hunk 38]
```diff
@@ -0,0 +1,2491 @@
...
+        tg_end = num_tiles - 1;
+    } else {
+        tg_start = u(log2_num_tiles);
+        tg_end = u(log2_num_tiles);
+    }
+
+	byte_alignment();
+	// Tile payload
+    int consumedBytes = (consumed_bits() + 7) / 8;
+    assert(consumedBytes > 0);
```
Comment (dabrain34) [inline vk_video_decoder/libs/NvVideoParser/src/VulkanAV1Decoder.cpp:2257]: I found a regression in CTS parser due to this line with an AV1 sample generated by VVS and RADV, see the test 
`./deqp-vk --deqp-vk-video-encode-dump=bitstream -n dEQP-VK.video.encode.av1.720x480_8le_420.i_15.lr_layered_dpb`

The generated IVF does not playback neither with VVS.

If I remove this line, I dont have the issue.

@vasn1k can you confirm the need or not of this line because it was not present in the initial work from Charlie to integrate the AV1 decoder in NVidiaParser, see https://github.com/Igalia/vk_video_samples/blob/435807a3f875d83aa00f5b4b0702dc156ba3ffa9/vk_video_decoder/libs/NvVideoParser/src/VulkanAV1Decoder.cpp#L2246

I created an issue https://github.com/KhronosGroup/Vulkan-Video-Samples/issues/96 and https://gitlab.khronos.org/Tracker/vk-gl-cts/-/issues/5978
